### PR TITLE
Bug 1864710 - Add UI for creating tabs

### PR DESCRIPTION
**Pull request by MozillaNoah**
https://github.com/mozilla-mobile/firefox-android/assets/87384386/dfc0abcd-ca2b-4d6c-8264-de246742bd4e

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1864710

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:217]: I'm going to file a follow-up after this lands to make this TextField a top-level component. 

The colors were all derived from the [figma](https://www.figma.com/file/ujdiReypSypu2GhRgzArea/Android-Components?type=design&node-id=3554-41479&mode=design&t=1tVWzjhdtyjyK59q-0)

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:335]: I'm debating adding logic to this preview to have the buttons also update the tab counter UI. (this is one of the downsides of removing `BrowserStore` as a parameter, as this update may have been handled for us.

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:217]: https://bugzilla.mozilla.org/show_bug.cgi?id=1871470

**Comment by MozillaNoah** [comment]
I've built a follow-up PR already for the plumbing of the Browser Store logic. It also has some logic and preview clean-up.

https://github.com/mozilla-mobile/firefox-android/pull/4931

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt
Comment (MatthewTighe) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:219]: This seems like a good idea! I honestly don't know much about handling `Keyboard` interactions correctly, so I'm just wondering if we're handling everything we need to from an accessibility standpoint

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:277]: Wondering if this should maybe be last, since it's the only conditional element

**Review by MatthewTighe — APPROVED** [review]
LGTM, happy to cycle again if you have any responses to my comments though

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:219]: For the purposes of this text field, we should be all set. When we stand up a `TextField` composable though, we'll need to create the correct API to capture/document all of the `Keyboard` interactions.

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:343]: I'm cool with this. We'll want to make `TabTools` depend on a boolean again then for `inactiveTabsEnabled`, instead of passing our `Settings` mega singleton as a dependency.

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:277]: My intent here was to mirror the ordering of the tab counts. Inactive tabs are really just a subset of normal tabs, so I wanted to keep them visually next to each other, if that makes sense.

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt
Comment (t-p-white) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:90]: I'm starting to wonder if we should re-consider this check - seems it's actually a bit of a legacy warning based on Java pre-Kotlin defaults & Compose

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:204]: nit: think these surrounding brackets are superfluous

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:99]: nit: think these surrounding brackets are superfluous

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt
Comment (t-p-white) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:90]: I think we need to reevaluate how we handle this rule. afaict we can change the rule to either:

- off completely for Compose functions
- all functions have the threshold increased

Comment (t-p-white) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/tabs/TabTools.kt:90]: Or can these be defaulted - then we can remove the suppression 🙂

**Pull request merged**
